File: src/main/java/org/lanternpowered/server/item/predicate/SimilarItemPredicate.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.item.predicate;

import org.lanternpowered.server.inventory.LanternItemStack;
import org.spongepowered.api.item.ItemType;
import org.spongepowered.api.item.inventory.ItemStack;
import org.spongepowered.api.item.inventory.ItemStackSnapshot;

/**
 * A {@link ItemPredicate} that tests whether the given
 * {@link ItemStack} is similar to the target stack.
 */
final class SimilarItemPredicate implements ItemPredicate {

    private final LanternItemStack itemStack;

    SimilarItemPredicate(LanternItemStack itemStack) {
        this.itemStack = itemStack;
    }

    @Override
    public boolean test(ItemStack stack) {
        return this.itemStack.similarTo(stack);
    }

    @Override
    public boolean test(ItemType type) {
        return this.itemStack.similarTo(ItemStack.of(type, 1));
    }

    @Override
    public boolean test(ItemStackSnapshot stack) {
        return this.itemStack.similarTo(stack);
    }
}

File: src/main/kotlin/org/lanternpowered/server/item/predicate/SimilarItemPredicate.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.item.predicate

import org.lanternpowered.api.item.ItemType
import org.lanternpowered.api.item.inventory.ExtendedItemStack
import org.lanternpowered.api.item.inventory.ItemStack
import org.lanternpowered.api.item.inventory.ItemStackSnapshot
import org.lanternpowered.api.item.inventory.itemStackOf

/**
 * A [ItemPredicate] that tests whether the given
 * [ItemStack] is similar to the target stack.
 */
internal class SimilarItemPredicate(private val itemStack: ExtendedItemStack) : ItemPredicate {

    override fun invoke(stack: ItemStack) = this.itemStack.isSimilarTo(stack)
    override fun invoke(type: ItemType) = this.itemStack.isSimilarTo(itemStackOf(type))
    override fun invoke(stack: ItemStackSnapshot) = this.itemStack.isSimilarTo(stack)
}
